File: my_claudecode_cpp/anthropic.cpp
#include "json.hpp"
#include <string>

using json = nlohmann::json;

json build_openai_request(const json& anthropic, const std::string& model_id) {
    json messages = json::array();
    if (anthropic.contains("system") && anthropic["system"].is_string()) {
        messages.push_back({{"role", "system"}, {"content", anthropic["system"]}});
    }
    if (anthropic.contains("messages") && anthropic["messages"].is_array()) {
        for (const auto& msg : anthropic["messages"]) {
            messages.push_back({{"role", msg["role"]}, {"content", msg["content"]}});
        }
    }
    return {
        {"model", model_id},
        {"messages", messages},
        {"stream", anthropic.value("stream", false)},
        {"max_tokens", anthropic.value("max_tokens", 4096)}
    };
}

json openai_to_anthropic(const json& openai, const std::string& model_id) {
    auto choice = openai["choices"][0];
    auto message = choice["message"];

    return {
        {"id", "msg_cpp_test"},
        {"type", "message"},
        {"role", "assistant"},
        {"model", model_id},
        {"content", {{{"type", "text"}, {"text", message["content"]}}}},
        {"stop_reason", "end_turn"},
        {"usage", {
            {"input_tokens", openai["usage"]["prompt_tokens"]},
            {"output_tokens", openai["usage"]["completion_tokens"]}
        }}
    };
}

File: my_claudecode_rust/src/anthropic.rs
use serde_json::{json, Value};
use uuid::Uuid;

pub fn build_openai_request(anthropic: Value, model_id: &str) -> Value {
    let mut messages = Vec::new();
    if let Some(system) = anthropic["system"].as_str() {
        messages.push(json!({"role": "system", "content": system}));
    }

    if let Some(msgs) = anthropic["messages"].as_array() {
        for msg in msgs {
            messages.push(json!({
                "role": msg["role"],
                "content": msg["content"]
            }));
        }
    }

    json!({
        "model": model_id,
        "messages": messages,
        "stream": anthropic["stream"].as_bool().unwrap_or(false),
        "max_tokens": anthropic["max_tokens"].as_i64().unwrap_or(4096),
        "temperature": anthropic["temperature"],
    })
}

pub fn openai_to_anthropic(openai: Value, model_id: &str) -> Value {
    let choice = &openai["choices"][0];
    let message = &choice["message"];

    json!({
        "id": format!("msg_rust_{}", Uuid::new_v4()),
        "type": "message",
        "role": "assistant",
        "model": model_id,
        "content": [{"type": "text", "text": message["content"]}],
        "stop_reason": if choice["finish_reason"] == "tool_calls" { "tool_use" } else { "end_turn" },
        "usage": {
            "input_tokens": openai["usage"]["prompt_tokens"],
            "output_tokens": openai["usage"]["completion_tokens"]
        }
    })
}
